### ISSUE TITLE
Provide Liquibase SLF4J bridge as an OSGi fragment

### ISSUE THREAD

**Issue by erwindl0** [mattbertolini/liquibase-slf4j#12]
In an OSGi runtime we also would like to bridge Liquibase 4.x's logging to SLF4J.

To get this working, Liquibase-SLF4J should be packaged as a fragment, with Liquibase core as the host bundle.
Then the Slf4JLogService can be discovered by Liquibase's ServiceLocator.

Cfr ongoing work and discussions:
* https://github.com/steinarb/liquibase-karaf-feature#notes-on-the-liquibase-logging-issues
* https://github.com/liquibase/liquibase/issues/2004 : don't want to wait for that one

I propose to prepare a PR for this?

**Comment by mattbertolini** [comment]
Hi. Thanks for reporting. I'm only vaguely familiar with OSGi and what I have been exposed to has left me a bit underwhelmed. I am pretty hesitant to add something I will need to maintain that I don't necessary understand. Based on that first link, it sounds like this fragment has been created and published. Is there some reason that the other library can't be used?

**Comment by erwindl0** [comment]
Thanks for the quick reaction.

I would prefer that the "real source" of the Liquibase-SLF4J bridge would support both basic Java runtimes and modular ones.
(Also, the Karaf feature project has not yet been updated to liquibase 4.x.)

But if you prefer I can either contact them to ask them to upgrade, or maintain my own fork independently?

**Comment by mattbertolini** [comment]
@erwindl0 I was able to generate this manifest with BND. Will this work?

```
Manifest-Version: 1.0
Ant-Version: Apache Ant 1.10.12
Automatic-Module-Name: com.mattbertolini.liquibase.logging.slf4j
Bnd-LastModified: 1643486539173
Bundle-ManifestVersion: 2
Bundle-Name: liquibase-slf4j
Bundle-SymbolicName: com.mattbertolini.liquibase.logging.slf4j
Bundle-Version: 4.1.0-SNAPSHOT
Created-By: 17 (Eclipse Adoptium)
Export-Package: com.mattbertolini.liquibase.logging.slf4j;version="4.1
 .0-SNAPSHOT"
Fragment-Host: org.liquibase.core;version=4.1.0
Implementation-Title: liquibase-slf4j
Implementation-Version: 4.1.0-SNAPSHOT
Import-Package: org.slf4j;version="[1.7,2)"
Originally-Created-By: 17+35 (Eclipse Adoptium)
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Specification-Title: liquibase-slf4j
Specification-Version: 4.1.0-SNAPSHOT
Tool: Bnd-6.1.0.202111221555
```

**Comment by mattbertolini** [comment]
I published a snapshot with the generated manifest. Let me know if it works for you.

https://oss.sonatype.org/content/repositories/snapshots/com/mattbertolini/liquibase-slf4j/4.1.0-SNAPSHOT/

**Comment by gmilani** [comment]
Hello guys,
I'm also trying to bridge Liquibase and SLF4J using Apache Karaf (4.2.15).
@mattbertolini I was able to install your SNAPSHOT, it is resolved and attached to liquibase-core (4.1.0), but for now still having "UnexpectedLiquibaseException: Cannot find default log service".

```
 47 │ Resolved │  60 │ 4.1.0.SNAPSHOT     │ liquibase-slf4j, Hosts: 247
247 │ Active   │  60 │ 4.1.0              │ liquibase-core, Fragments: 47
```

```
liquibase-slf4j (47)
--------------------
Ant-Version = Apache Ant 1.10.10
Automatic-Module-Name = com.mattbertolini.liquibase.logging.slf4j
Bnd-LastModified = 1643552376881
Created-By = 11.0.14 (Eclipse Adoptium)
Fragment-Host = org.liquibase.core;bundle-version="[4.1.0,5)"
Implementation-Title = liquibase-slf4j
Implementation-Version = 4.1.0-SNAPSHOT
Manifest-Version = 1.0
Originally-Created-By = 11.0.14+9 (Eclipse Adoptium)
Specification-Title = liquibase-slf4j
Specification-Version = 4.1.0-SNAPSHOT
Tool = Bnd-6.1.0.202111221555

Bundle-ManifestVersion = 2
Bundle-Name = liquibase-slf4j
Bundle-SymbolicName = com.mattbertolini.liquibase.logging.slf4j
Bundle-Version = 4.1.0.SNAPSHOT

Require-Capability = 
	osgi.ee;filter:=(&(osgi.ee=JavaSE)(version=1.8))

Export-Package = 
	com.mattbertolini.liquibase.logging.slf4j;version=4.1.0.SNAPSHOT
Import-Package = 
	org.slf4j;version="[1.7,2)"
```

I'll let you know if I find more information about this issue.

**Comment by mattbertolini** [comment]
Thanks for the report. I don't know too much about OSGi so any assistance would be appreciated. Do you have a stacktrace for the error? That will help pinpoint where in Liquibase its having issues. 

The "Cannot find default log service" usually means something is trying to load the service before the scope is fully active in Liquibase. I also know that Liquibase frequently breaks OSGi support. I think they just fixed something else in 4.7.1 so I would try an updated version. 

Lastly, I don't have an example project using OSGi so if you one or have pointers on how to set one up, that would be greatly appreciated.

**Comment by mattbertolini** [comment]
I've published another snapshot to maven central snapshots. I've added the following to my manifest:

```
Provide-Capability: osgi.serviceloader; osgi.serviceloader=liquibase.logging.LogService
Require-Capability: osgi.serviceloader; osgi.serviceloader=liquibase.logging.LogService
```

I realized that I wasn't exposing the JDK `ServiceLoader` info and after reading a blog post or two on OSGi I thought it was worth a shot.

**Comment by gmilani** [comment]
Hi @mattbertolini ,
Unfortunately I still have the same issue. To help you (or anybody else who could help), I prepared a basic Karaf application.
It is supposed to execute a migration script on a HSQL database when the datasource is created. The migrator is executed and the exception I mentioned before happens.

Of course, I were not further in the process and there will maybe an other issue with Liquibase during the migration once the logging issue is solved, so don't be surprised if the script isn't applied to the database.

https://github.com/gmilani/karaf-liquibase-logging

Don't hesitate to ask if I can help you, having your library running on OSGi would be a really valuable achievement for me.
Thank you for the job you already made !

**Comment by mattbertolini** [comment]
Thanks @gmilani. This is very helpful.

Are we sure that Liquibase works without the slfj4j jar? I ask because even if slfj4j isn't found from the bundle the default logger should be found and it is not. I am not sure the version of Liquibase used (4.4.x) is working with OSGi. Getting it to work without the slf4j logging bridge jar is the first step to knowing what's going on.

**Comment by erwindl0** [comment]
Hi all,

 

Sorry I have not been active in helping out on this.

 

But just a rapid confirmation:

 

When you run liquibase(4.4.3 in my case) in an OSGi container (Equinox in my case), it does work.

But it uses the std Java logger which I am unable to configure in my environment.

There is no dependency on SLF4j in Liquibase itself (as can be checked in their MANIFEST).

 

The result of mixed logging (Slf4j for everything except Liquibase; java.util.loggin for Liquibase) is similar to below:

(LiquibaseChangeDeployer is my code to react on an event in the application to then trigger Liquibase)

 

2022-02-25 11:11:43 TRACE [ Start Level: Equinox Container: 71916ac5-8daf-4a91-9552-1f8e6fc17534     ] checkDatabaseChanges() - entry - PU master [LiquibaseChangeDeployer:193]

2022-02-25 11:11:43 TRACE [ Start Level: Equinox Container: 71916ac5-8daf-4a91-9552-1f8e6fc17534     ] deployDatabaseChanges() - entry : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:69]

Feb 25, 2022 11:11:43 AM liquibase.servicelocator

INFO: Cannot load service: liquibase.license.LicenseService: Provider liquibase.license.pro.DaticalTrueLicenseService could not be instantiated

2022-02-25 11:11:43 TRACE [ Start Level: Equinox Container: 71916ac5-8daf-4a91-9552-1f8e6fc17534     ] checkPreconditions() - entry : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:126]

2022-02-25 11:11:43 TRACE [ Start Level: Equinox Container: 71916ac5-8daf-4a91-9552-1f8e6fc17534     ] checkPreconditions() - exit : changelog pas_xr_master_changelog.xml for PU master - return true [LiquibaseChangeDeployer:173]

Feb 25, 2022 11:11:43 AM liquibase.lockservice

INFO: Successfully acquired change log lock

Feb 25, 2022 11:11:44 AM liquibase.changelog

INFO: Reading from `xr_portal_2.0_master`.DATABASECHANGELOG

Feb 25, 2022 11:11:44 AM liquibase.lockservice

INFO: Successfully released change log lock

2022-02-25 11:11:44 INFO  [ Start Level: Equinox Container: 71916ac5-8daf-4a91-9552-1f8e6fc17534     ] Deployed database changes pas_xr_master_changelog.xml in PU master [LiquibaseChangeDeployer:91]

2022-02-25 11:11:44 TRACE [ Start Level: Equinox Container: 71916ac5-8daf-4a91-9552-1f8e6fc17534     ] deployDatabaseChanges() - exit : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:111]

2022-02-25 11:11:44 TRACE [ Start Level: Equinox Container: 71916ac5-8daf-4a91-9552-1f8e6fc17534     ] checkDatabaseChanges() - exit - PU master [LiquibaseChangeDeployer:205]

 

 

With a correctly built liquibase-slf4j fragment, we get a single integrated slf4j-based log as below.

(I have defined the manifest manually though, I'll try later today with the one you generated with BND) :

 

2022-02-25 11:09:06 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] checkDatabaseChanges() - entry - PU master [LiquibaseChangeDeployer:193]

2022-02-25 11:09:06 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] deployDatabaseChanges() - entry : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:69]

2022-02-25 11:09:06 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Cannot load service: liquibase.license.LicenseService: Provider liquibase.license.pro.DaticalTrueLicenseService could not be instantiated [StandardServiceLocator:153]

2022-02-25 11:09:06 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] checkPreconditions() - entry : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:126]

2022-02-25 11:09:06 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] checkPreconditions() - exit : changelog pas_xr_master_changelog.xml for PU master - return true [LiquibaseChangeDeployer:173]

2022-02-25 11:09:07 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Successfully acquired change log lock [StandardLockService:153]

2022-02-25 11:09:07 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Reading from `xr_portal_2.0_master`.DATABASECHANGELOG [StandardChangeLogHistoryService:153]

2022-02-25 11:09:08 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Successfully released change log lock [StandardLockService:153]

2022-02-25 11:09:08 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Deployed database changes pas_xr_master_changelog.xml in PU master [LiquibaseChangeDeployer:91]

2022-02-25 11:09:08 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] deployDatabaseChanges() - exit : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:111]

2022-02-25 11:09:08 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] checkDatabaseChanges() - exit - PU master [LiquibaseChangeDeployer:205]

 

 

regards

erwin

**Comment by erwindl0** [comment]
Hi all,

Sorry I have not been active in helping out on this.

But just a rapid confirmation:

When you run liquibase(4.4.3 in my case) in an OSGi container (Equinox in my case), it does work.
But it uses the std Java logger which I am unable to configure in my environment.
There is no dependency on SLF4j in Liquibase itself (as can be checked in their MANIFEST).

The result of mixed logging (Slf4j for everything except Liquibase; java.util.loggin for Liquibase) is similar to below:
(LiquibaseChangeDeployer is my code to react on an event in the application to then trigger Liquibase)

 
![image](https://user-images.githubusercontent.com/6461921/155700601-4580afb0-3a39-4020-b12e-9ee2135bd570.png)
 

With a correctly built liquibase-slf4j fragment, we get a single integrated slf4j-based log as below.
(I have defined the manifest manually though, I'll try later today with the one you generated with BND) :

```
2022-02-25 11:09:06 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] checkDatabaseChanges() - entry - PU master [LiquibaseChangeDeployer:193]
2022-02-25 11:09:06 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] deployDatabaseChanges() - entry : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:69]
2022-02-25 11:09:06 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Cannot load service: liquibase.license.LicenseService: Provider liquibase.license.pro.DaticalTrueLicenseService could not be instantiated [StandardServiceLocator:153]
2022-02-25 11:09:06 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] checkPreconditions() - entry : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:126]
2022-02-25 11:09:06 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] checkPreconditions() - exit : changelog pas_xr_master_changelog.xml for PU master - return true [LiquibaseChangeDeployer:173]
2022-02-25 11:09:07 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Successfully acquired change log lock [StandardLockService:153]
2022-02-25 11:09:07 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Reading from `xr_portal_2.0_master`.DATABASECHANGELOG [StandardChangeLogHistoryService:153]
2022-02-25 11:09:08 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Successfully released change log lock [StandardLockService:153]
2022-02-25 11:09:08 INFO  [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] Deployed database changes pas_xr_master_changelog.xml in PU master [LiquibaseChangeDeployer:91]
2022-02-25 11:09:08 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] deployDatabaseChanges() - exit : changelog pas_xr_master_changelog.xml for PU master [LiquibaseChangeDeployer:111]
2022-02-25 11:09:08 TRACE [ Start Level: Equinox Container: cb0a510a-ede8-4c04-a6b6-d34f513807e7     ] checkDatabaseChanges() - exit - PU master [LiquibaseChangeDeployer:205]
```

regards

erwin

**Comment by erwindl0** [comment]
I just checked with your snapshot jar and in my osgi application/platform it does seem to work.
But as mentioned I'm still on Liquibase 4.4.3.

I.e. I get the right integrated logging.
And host&fragment are correctly associated:

id	State       Bundle
77	RESOLVED    com.mattbertolini.liquibase.logging.slf4j_4.1.0.SNAPSHOT
	            Master=236
236	ACTIVE      org.liquibase.core_4.4.3
	            Fragments=77

**Comment by mattbertolini** [comment]
That good to hear @erwindl0 . Thanks for taking a look.

Based on the available information, I'm inclined to believe the manifest is correct and there is something else wrong with either Liquibase or the example. 

@gmilani I tried running the provided example without liquibase-slf4j to see if it didn't throw the `UnexpectedLiquibaseException`. Unfortunately it continued to throw the same error so I have to believe that there is something else wrong here. Liquibase should log to standard out by default so this exception could only appear if Liquibase itself couldn't find its own loggers (which are in the same bundle). 

I noticed that the example doesn't use the `OSGiResourceAccessor` provided by Liquibase. Is there a particular reason for this? The original code example provided above seemed to use it. Also, is there a particular reason you are creating the `Database` object manually rather than passing the `JdbcConnection` to Liquibase directly. I was thinking simplifying the example might improve things, like this:

```java
private void prepare(Connection connection) throws LiquibaseException {
        OSGiResourceAccessor resourceAccessor = new OSGiResourceAccessor(bundle);
        try (Liquibase liquibase = new Liquibase("migrations.xml", resourceAccessor, new JdbcConnection(connection))) {
            liquibase.update(new Contexts());
        }
    }
```

Given @erwindl0's information, I might move ahead with publishing what I have on branch gen-osgi-bundle-bnd. Let me know your thoughts. Thanks.
